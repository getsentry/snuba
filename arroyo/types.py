from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Generic, Optional, TypeVar


@dataclass(order=True, unsafe_hash=True)
class Topic:
    __slots__ = ["name"]

    name: str

    def __contains__(self, partition: Partition) -> bool:
        return partition.topic == self


@dataclass(order=True, unsafe_hash=True)
class Partition:
    __slots__ = ["topic", "index"]

    topic: Topic
    index: int


TPayload = TypeVar("TPayload")


@dataclass(unsafe_hash=True)
class Message(Generic[TPayload]):
    """
    Represents a single message within a partition.
    """

    __slots__ = ["partition", "offset", "payload", "timestamp", "next_offset"]

    partition: Partition
    offset: int
    payload: TPayload
    timestamp: datetime
    next_offset: int

    def __init__(
        self,
        partition: Partition,
        offset: int,
        payload: TPayload,
        timestamp: datetime,
        next_offset: Optional[int] = None,
    ) -> None:
        if next_offset is None:
            next_offset = offset + 1

        self.partition = partition
        self.offset = offset
        self.payload = payload
        self.timestamp = timestamp
        self.next_offset = next_offset

    def __repr__(self) -> str:
        # XXX: Field values can't be excluded from ``__repr__`` with
        # ``dataclasses.field(repr=False)`` as this class is defined with
        # ``__slots__`` for performance reasons. The class variable names
        # would conflict with the instance slot names, causing an error.
        return f"{type(self).__name__}(partition={self.partition!r}, offset={self.offset!r})"
