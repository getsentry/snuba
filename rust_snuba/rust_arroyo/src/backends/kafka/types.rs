use rdkafka::message::OwnedHeaders;

#[derive(Clone)]
pub struct KafkaPayload {
    pub key: Option<Vec<u8>>,
    pub headers: Option<OwnedHeaders>,
    pub payload: Option<Vec<u8>>,
}
