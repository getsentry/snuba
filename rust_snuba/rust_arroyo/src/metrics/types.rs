use core::fmt::{self, Display};
use std::time::Duration;

/// The Type of a Metric.
///
/// Counters, Gauges and Distributions are supported,
/// with more types to be added later.
#[non_exhaustive]
#[derive(Debug)]
pub enum MetricType {
    /// A counter metric, using the StatsD `c` type.
    Counter,
    /// A gauge metric, using the StatsD `g` type.
    Gauge,
    /// A timer metric, using the StatsD `ms` type.
    Timer,
    // Distribution,
    // Meter,
    // Histogram,
    // Set,
}

impl MetricType {
    /// Returns the StatsD metrics type.
    pub fn as_str(&self) -> &str {
        match self {
            MetricType::Counter => "c",
            MetricType::Gauge => "g",
            MetricType::Timer => "ms",
        }
    }
}

impl Display for MetricType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// A Metric Value.
///
/// This supports various numeric values for now, but might gain support for
/// `Duration` and other types later on.
#[non_exhaustive]
#[derive(Debug)]
pub enum MetricValue {
    /// A signed value.
    I64(i64),
    /// An unsigned value.
    U64(u64),
    /// A floating-point value.
    F64(f64),
    /// A [`Duration`] value.
    Duration(Duration),
}

impl Display for MetricValue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MetricValue::I64(v) => v.fmt(f),
            MetricValue::U64(v) => v.fmt(f),
            MetricValue::F64(v) => v.fmt(f),
            MetricValue::Duration(d) => d.as_millis().fmt(f),
        }
    }
}

macro_rules! into_metric_value {
    ($($from:ident),+ => $variant:ident) => {
        $(
            impl From<$from> for MetricValue {
                #[inline(always)]
                fn from(f: $from) -> Self {
                    Self::$variant(f.into())
                }
            }
        )+
    };
}

into_metric_value!(i8, i16, i32, i64 => I64);
into_metric_value!(u8, u16, u32, u64 => U64);
into_metric_value!(f32, f64 => F64);
into_metric_value!(Duration => Duration);

/// An alias for a list of Metric tags.
pub type MetricTags<'a> = &'a [(Option<&'a dyn Display>, &'a dyn Display)];

/// A fully types Metric.
///
/// Most importantly, the metric has a [`ty`](MetricType), a `key` and a [`value`](MetricValue).
/// It can also have a list of [`tags`](MetricTags).
///
/// This struct might change in the future, and one should construct it via
/// the [`metric!`](crate::metric) macro instead.
pub struct Metric<'a> {
    /// The key, or name, of the metric.
    pub key: &'a dyn Display,
    /// The type of metric.
    pub ty: MetricType,

    /// A list of tags for this metric.
    pub tags: MetricTags<'a>,
    /// The metrics value.
    pub value: MetricValue,

    #[doc(hidden)]
    pub __private: (),
}
