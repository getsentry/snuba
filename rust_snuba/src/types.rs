use std::cmp::min;
use std::collections::BTreeMap;

use crate::strategies::clickhouse::batch::HttpBatch;

use chrono::{DateTime, Utc};
use rust_arroyo::backends::kafka::types::KafkaPayload;
use rust_arroyo::timer;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq)]
pub struct CommitLogEntry {
    pub offset: u64,
    pub orig_message_ts: DateTime<Utc>,
    pub received_p99: Vec<DateTime<Utc>>,
}

#[derive(Default, Debug, Clone)]
pub struct CommitLogOffsets(pub BTreeMap<u16, CommitLogEntry>);

impl CommitLogOffsets {
    fn merge(&mut self, other: CommitLogOffsets) {
        for (partition, other_entry) in other.0 {
            self.0
                .entry(partition)
                .and_modify(|entry| {
                    entry.offset = other_entry.offset;
                    entry.orig_message_ts = other_entry.orig_message_ts;
                    entry.received_p99.extend(&other_entry.received_p99);
                })
                .or_insert(other_entry);
        }
    }
}

#[derive(Clone, Debug, PartialEq, Default)]
pub struct CogsData {
    pub data: BTreeMap<String, u64>, // app_feature: bytes_len
}

impl CogsData {
    fn merge(&mut self, other: CogsData) {
        for (k, v) in other.data {
            self.data
                .entry(k)
                .and_modify(|curr| *curr += v)
                .or_insert(v);
        }
    }
}

#[derive(Debug, Clone)]
struct LatencyRecorder {
    sum_timestamps: f64,
    earliest_timestamp: u64,
    num_values: usize,
}

impl Default for LatencyRecorder {
    fn default() -> Self {
        LatencyRecorder {
            sum_timestamps: 0.0,
            earliest_timestamp: u64::MAX,
            num_values: 0,
        }
    }
}

impl From<DateTime<Utc>> for LatencyRecorder {
    fn from(value: DateTime<Utc>) -> Self {
        let value = value.timestamp_millis();
        LatencyRecorder {
            sum_timestamps: value as f64,
            earliest_timestamp: value as u64,
            num_values: 1,
        }
    }
}

impl LatencyRecorder {
    fn merge(&mut self, other: Self) {
        self.sum_timestamps += other.sum_timestamps;
        self.earliest_timestamp = min(self.earliest_timestamp, other.earliest_timestamp);
        self.num_values += other.num_values;
    }

    fn max_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        write_time.saturating_sub(self.earliest_timestamp)
    }

    fn avg_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        (write_time as f64 - (self.sum_timestamps / self.num_values as f64)) as u64
    }

    fn send_metric(&self, write_time: DateTime<Utc>, metric_name: &str) {
        if self.num_values == 0 {
            return;
        }

        timer!(
            format_args!("insertions.max_{}_ms", metric_name),
            self.max_value_ms(write_time)
        );

        timer!(
            format_args!("insertions.{}_ms", metric_name),
            self.avg_value_ms(write_time),
        );
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReplacementData {
    pub key: Vec<u8>, // Project id
    // Replacement message bytes. Newline-delimited series of json payloads
    pub value: Vec<u8>,
}

impl From<ReplacementData> for KafkaPayload {
    fn from(value: ReplacementData) -> KafkaPayload {
        KafkaPayload::new(Some(value.key), None, Some(value.value))
    }
}

impl From<KafkaPayload> for ReplacementData {
    fn from(value: KafkaPayload) -> ReplacementData {
        ReplacementData {
            key: value.key().unwrap().clone(),
            value: value.payload().unwrap().clone(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum InsertOrReplacement<T> {
    Insert(T),
    Replacement(ReplacementData),
}

/// The return value of message processors.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct InsertBatch {
    pub rows: RowData,
    pub origin_timestamp: Option<DateTime<Utc>>,
    pub sentry_received_timestamp: Option<DateTime<Utc>>,
    pub cogs_data: Option<CogsData>,
}

impl InsertBatch {
    pub fn from_rows<T>(
        rows: impl IntoIterator<Item = T>,
        origin_timestamp: Option<DateTime<Utc>>,
    ) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let rows = RowData::from_rows(rows)?;
        Ok(Self {
            rows,
            origin_timestamp,
            sentry_received_timestamp: None,
            cogs_data: None,
        })
    }

    /// In case the processing function wants to skip the message, we return an empty batch.
    /// But instead of having the caller send an empty batch, lets make an explicit api for
    /// skipping. This way we can change the implementation later if we want to. Skipping ensures
    /// that the message is committed but not processed.
    pub fn skip() -> Self {
        Self::default()
    }
}

#[derive(Clone, Debug, Default)]
pub struct BytesInsertBatch<R> {
    rows: R,

    /// when the message was inserted into the snuba topic
    ///
    /// In Python this aggregate value is not explicitly tracked on BytesInsertBatch, as the batch
    /// type contains a list of the original kafka messages which all contain the individual
    /// timestamp values in metadata.
    message_timestamp: LatencyRecorder,

    /// when the event was received by Relay
    origin_timestamp: LatencyRecorder,

    /// when it was received by the ingest consumer in Sentry
    ///
    /// May not be recorded for some datasets. This is specifically used for metrics datasets, where
    /// this represents the latency from ingest-metrics (i.e. before the metrics indexer) to
    /// insertion into clickhouse
    sentry_received_timestamp: LatencyRecorder,

    // For each partition we store the offset and timestamp to be produced to the commit log
    commit_log_offsets: CommitLogOffsets,

    cogs_data: CogsData,
}

impl<R> BytesInsertBatch<R> {
    pub fn new(
        rows: R,
        message_timestamp: Option<DateTime<Utc>>,
        origin_timestamp: Option<DateTime<Utc>>,
        sentry_received_timestamp: Option<DateTime<Utc>>,
        commit_log_offsets: CommitLogOffsets,
        cogs_data: CogsData,
    ) -> Self {
        BytesInsertBatch {
            rows,
            message_timestamp: message_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            origin_timestamp: origin_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            sentry_received_timestamp: sentry_received_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            commit_log_offsets,
            cogs_data,
        }
    }

    pub fn commit_log_offsets(&self) -> &CommitLogOffsets {
        &self.commit_log_offsets
    }

    pub fn cogs_data(&self) -> &CogsData {
        &self.cogs_data
    }

    pub fn take(self) -> (R, BytesInsertBatch<()>) {
        let new = BytesInsertBatch {
            rows: (),
            message_timestamp: self.message_timestamp,
            origin_timestamp: self.origin_timestamp,
            sentry_received_timestamp: self.sentry_received_timestamp,
            commit_log_offsets: self.commit_log_offsets,
            cogs_data: self.cogs_data,
        };

        (self.rows, new)
    }

    pub fn record_message_latency(&self) {
        let write_time = Utc::now();

        self.message_timestamp.send_metric(write_time, "latency");
        self.origin_timestamp
            .send_metric(write_time, "end_to_end_latency");
        self.sentry_received_timestamp
            .send_metric(write_time, "sentry_received_latency");
    }
}

impl BytesInsertBatch<RowData> {
    pub fn len(&self) -> usize {
        self.rows.num_rows
    }
}

impl BytesInsertBatch<HttpBatch> {
    pub fn merge(mut self, other: BytesInsertBatch<RowData>) -> Self {
        self.rows
            .write_rows(&other.rows)
            .expect("failed to write rows to channel");
        self.commit_log_offsets.merge(other.commit_log_offsets);
        self.message_timestamp.merge(other.message_timestamp);
        self.origin_timestamp.merge(other.origin_timestamp);
        self.sentry_received_timestamp
            .merge(other.sentry_received_timestamp);
        self.cogs_data.merge(other.cogs_data);
        self
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
pub struct RowData {
    pub encoded_rows: Vec<u8>,
    pub num_rows: usize,
}

impl RowData {
    pub fn from_rows<T>(rows: impl IntoIterator<Item = T>) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            serde_json::to_writer(&mut encoded_rows, &row)?;
            debug_assert!(encoded_rows.ends_with(b"}"));
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        Ok(RowData {
            num_rows,
            encoded_rows,
        })
    }

    pub fn from_encoded_rows(rows: Vec<Vec<u8>>) -> Self {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            encoded_rows.extend_from_slice(&row);
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        RowData {
            encoded_rows,
            num_rows,
        }
    }

    pub fn into_encoded_rows(self) -> Vec<u8> {
        self.encoded_rows
    }
}

#[derive(Clone, Debug)]
pub struct KafkaMessageMetadata {
    pub partition: u16,
    pub offset: u64,
    pub timestamp: DateTime<Utc>,
}

#[cfg(test)]
mod tests {
    use chrono::{TimeZone, Utc};

    use super::*;

    #[test]
    fn test_latency_recorder_basic() {
        // emulate how Reduce works when creating a batch for inserting into clickhouse, because
        // that's where we use it in practice. Reduce starts with a default() accumulator
        let mut accumulator = LatencyRecorder::default();

        // then there's one row being merged into the current batch
        let now = Utc.timestamp_opt(61, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // then another row
        let now = Utc.timestamp_opt(63, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // finally, when flushing the batch, we record the metric. let's see if LatencyRecorder did
        // the math right:
        assert_eq!(accumulator.num_values, 2);
        let now = Utc.timestamp_opt(65, 0).unwrap();
        assert_eq!(accumulator.max_value_ms(now), 4000);
        assert_eq!(accumulator.avg_value_ms(now), 3000);
    }
}
