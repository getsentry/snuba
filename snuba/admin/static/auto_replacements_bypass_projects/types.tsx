import { ReactNode } from "react";

type AutoReplacementsBypassProjectsData = {
  projectID: number;
  expiry: string;
};

export { AutoReplacementsBypassProjectsData };
