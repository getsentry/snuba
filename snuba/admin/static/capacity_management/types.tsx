import { ReactNode } from "react";

type AllocationPolicy = {
  policy_name: string;
  configs: AllocationPolicyConfig[];
  optional_config_definitions: AllocationPolicyOptionalConfigDefinition[];
};

type AllocationPolicyConfig = {
  name: string;
  value: string;
  description: string;
  type: string;
  params: object;
};

type AllocationPolicyConfigParams = {
  name: string;
  type: string;
};

type AllocationPolicyOptionalConfigDefinition = {
  name: string;
  type: string;
  default: string;
  description: string;
  params: AllocationPolicyConfigParams[];
};

type RowData = {
  name: ReactNode;
  params: ReactNode;
  value: ReactNode;
  description: ReactNode;
  type: ReactNode;
  edit: ReactNode;
};

export {
  AllocationPolicy,
  AllocationPolicyConfig,
  AllocationPolicyOptionalConfigDefinition,
  AllocationPolicyConfigParams,
  RowData,
};
