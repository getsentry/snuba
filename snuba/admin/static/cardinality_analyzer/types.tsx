type QueryResultColumnMetadata = [string];
type QueryResultRow = [string];

type CardinalityQueryRequest= {
  sql: string;
};

type CardinalityQueryResult = {
  input_query: string;
  timestamp: number;
  column_names: QueryResultColumnMetadata;
  rows: [QueryResultRow];
  error?: string;
};

type PredefinedQuery = {
  name: string;
  sql: string;
  description: string;
};

export { CardinalityQueryRequest, CardinalityQueryResult, PredefinedQuery };
