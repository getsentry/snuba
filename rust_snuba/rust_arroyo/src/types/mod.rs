use chrono::{DateTime, Utc};
use std::any::type_name;
use std::cmp::Eq;
use std::fmt;
use std::hash::Hash;

#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Topic {
    pub name: String,
}

impl fmt::Display for Topic {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Topic({})", self.name)
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct Partition {
    // TODO: Make this a reference to 'static Topic.
    pub topic: Topic,
    pub index: u16,
}

impl fmt::Display for Partition {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Partition({} topic={})", self.index, &self.topic)
    }
}

#[derive(PartialEq)]
pub enum TopicOrPartition {
    Topic(Topic),
    Partition(Partition),
}

#[derive(Clone, Debug, PartialEq)]
pub struct Position {
    pub offset: u64,
    pub timestamp: DateTime<Utc>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Message<T: Clone> {
    pub partition: Partition,
    pub offset: u64,
    pub payload: T,
    pub timestamp: DateTime<Utc>,
}

impl<T: Clone> Message<T> {
    pub fn new(partition: Partition, offset: u64, payload: T, timestamp: DateTime<Utc>) -> Self {
        Self {
            partition,
            offset,
            payload,
            timestamp,
        }
    }
    pub fn next_offset(&self) -> u64 {
        self.offset + 1
    }
}

impl<T: Clone> fmt::Display for Message<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Message<{}>(partition={}), offset={}",
            type_name::<T>(),
            &self.partition,
            &self.offset
        )
    }
}

#[cfg(test)]
mod tests {
    use super::{Message, Partition, Topic};
    use chrono::Utc;
    use std::collections::HashMap;

    #[test]
    fn message() {
        let now = Utc::now();
        let topic = Topic {
            name: "test".to_string(),
        };
        let part = Partition { topic, index: 10 };
        let message = Message::new(part, 10, "payload".to_string(), now);

        assert_eq!(message.partition.topic.name, "test");
        assert_eq!(message.partition.index, 10);
        assert_eq!(message.offset, 10);
        assert_eq!(message.payload, "payload");
        assert_eq!(message.timestamp, now);
        assert_eq!(message.next_offset(), 11)
    }

    #[test]
    fn fmt_display() {
        let now = Utc::now();
        let part = Partition {
            topic: Topic {
                name: "test".to_string(),
            },
            index: 10,
        };
        let message = Message::new(part, 10, "payload".to_string(), now);

        assert_eq!(
            message.to_string(),
            "Message<alloc::string::String>(partition=Partition(10 topic=Topic(test))), offset=10"
        )
    }

    #[test]
    fn test_eq() {
        let a = Topic {
            name: "test".to_string(),
        };
        let b = Topic {
            name: "test".to_string(),
        };
        assert!(a == b);

        let c = Topic {
            name: "test2".to_string(),
        };
        assert!(a != c);
    }

    #[test]
    fn test_hash() {
        let mut content = HashMap::new();
        content.insert(
            Topic {
                name: "test".to_string(),
            },
            "test_value".to_string(),
        );

        let b = Topic {
            name: "test".to_string(),
        };
        let c = content.get(&b).unwrap();
        assert_eq!(&"test_value".to_string(), c);
    }

    #[test]
    fn test_clone() {
        let topic = Topic {
            name: "test".to_string(),
        };
        let part = Partition { topic, index: 10 };

        let part2 = part.clone();
        assert_eq!(part, part2);
        assert_ne!(&part as *const Partition, &part2 as *const Partition);

        let now = Utc::now();
        let message = Message::new(part, 10, "payload".to_string(), now);
        let message2 = message.clone();

        assert_eq!(message, message2);
        assert_ne!(
            &message as *const Message<String>,
            &message2 as *const Message<String>
        );
    }
}
