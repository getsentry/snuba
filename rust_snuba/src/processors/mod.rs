mod functions;
mod profiles;
mod querylog;
mod spans;
mod utils;

use crate::types::{KafkaMessageMetadata, RowData};
use rust_arroyo::backends::kafka::types::KafkaPayload;

type ProcessingFunction = fn(KafkaPayload, KafkaMessageMetadata) -> anyhow::Result<RowData>;

pub fn get_processing_function(name: &str) -> Option<ProcessingFunction> {
    match name {
        "FunctionsMessageProcessor" => Some(functions::process_message),
        "ProfilesMessageProcessor" => Some(profiles::process_message),
        "QuerylogProcessor" => Some(querylog::process_message),
        "SpansMessageProcessor" => Some(spans::process_message),
        _ => None,
    }
}
