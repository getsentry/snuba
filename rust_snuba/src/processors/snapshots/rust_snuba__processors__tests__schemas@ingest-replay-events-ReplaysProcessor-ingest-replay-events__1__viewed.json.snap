---
source: src/processors/mod.rs
description: "{\n  \"type\": \"replay_event\",\n  \"start_time\": 1712009295.0,\n  \"replay_id\": \"834f314caae54030a1b0dc52b202f24a\",\n  \"project_id\": 1,\n  \"retention_days\": 90,\n  \"payload\": [\n    123, 10, 32, 32, 32, 32, 34, 116, 121, 112, 101, 34, 58, 32, 34, 114, 101,\n    112, 108, 97, 121, 95, 118, 105, 101, 119, 101, 100, 34, 44, 10, 32, 32, 32,\n    32, 34, 118, 105, 101, 119, 101, 100, 95, 98, 121, 95, 105, 100, 34, 58, 32,\n    50, 51, 56, 50, 51, 54, 50, 51, 50, 52, 50, 44, 10, 32, 32, 32, 32, 34, 116,\n    105, 109, 101, 115, 116, 97, 109, 112, 34, 58, 32, 49, 55, 49, 50, 48, 48,\n    57, 50, 57, 53, 10, 125\n  ]\n}\n"
expression: snapshot_payload
---
[
  {
    "browser_name": "",
    "browser_version": "",
    "click_alt": "",
    "click_aria_label": "",
    "click_class": [],
    "click_component_name": "",
    "click_id": "",
    "click_is_dead": 0,
    "click_is_rage": 0,
    "click_node_id": 0,
    "click_role": "",
    "click_tag": "",
    "click_testid": "",
    "click_text": "",
    "click_title": "",
    "debug_id": "00000000-0000-0000-0000-000000000000",
    "device_brand": "",
    "device_family": "",
    "device_model": "",
    "device_name": "",
    "dist": "",
    "environment": "",
    "error_id": "00000000-0000-0000-0000-000000000000",
    "error_ids": [],
    "error_sample_rate": -1.0,
    "event_hash": "<event UUID>",
    "fatal_id": "00000000-0000-0000-0000-000000000000",
    "info_id": "00000000-0000-0000-0000-000000000000",
    "ip_address_v4": null,
    "ip_address_v6": null,
    "is_archived": 0,
    "offset": 1,
    "os_name": "",
    "os_version": "",
    "partition": 0,
    "platform": "javascript",
    "project_id": 1,
    "release": "",
    "replay_id": "834f314c-aae5-4030-a1b0-dc52b202f24a",
    "replay_start_timestamp": null,
    "replay_type": "",
    "retention_days": 90,
    "sdk_name": "",
    "sdk_version": "",
    "segment_id": null,
    "session_sample_rate": -1.0,
    "tags.key": [],
    "tags.value": [],
    "timestamp": 1712009295,
    "title": null,
    "trace_ids": [],
    "urls": [],
    "user": "",
    "user_email": "",
    "user_id": "",
    "user_name": "",
    "viewed_by_id": 23823623242,
    "warning_id": "00000000-0000-0000-0000-000000000000"
  }
]
